File: projects/13-live-ingest/src/fmp4.rs
//! V3 — Live fMP4 repackaging: rewrap H.264/AAC into CMAF, no re-encode.
//!
//! The codecs arriving over RTMP (H.264 in AVCC, AAC) are already `<video>`-playable —
//! the job is not to transcode them but to **remux**: rebuild a CMAF **init segment**
//! (`ftyp` + `moov` carrying the codec setup, zero samples) and a running series of
//! **fragments** (`moof` + `mdat`) on a monotonic MP4 timeline, cutting **parts**
//! (~200 ms) on demand and **segments** on IDR keyframes. This overlaps project 11's
//! `segment.rs` box-writing — reuse it. What's new is doing it **live**: the timeline
//! comes from RTMP message timestamps (32-bit, wrapping), not a finished sample table,
//! and you only ever hold the current fragment's samples (the memory bound).

use bytes::Bytes;

use crate::error::AppError;

/// Codec setup extracted from the first RTMP audio/video tags (V2), enough to write an
/// init segment. The AVC decoder config is the `avcC` (SPS/PPS); the AAC config is the
/// AudioSpecificConfig. This is the *setup*, carried in `moov` — not per-frame media.
#[derive(Debug, Clone)]
pub struct CodecConfig {
    /// The `avcC` box payload (SPS/PPS), as delivered in the AVC sequence header.
    pub avc_decoder_config: Bytes,
    /// The AAC AudioSpecificConfig, as delivered in the AAC sequence header.
    pub aac_audio_specific_config: Bytes,
    pub width: u16,
    pub height: u16,
    /// MP4 movie timescale (e.g. 90_000 for video). RTMP ms timestamps map into this.
    pub timescale: u32,
}

/// One access unit (a coded video frame or audio frame) to place in a fragment.
#[derive(Debug, Clone)]
pub struct Sample {
    /// The elementary bytes: length-prefixed NALUs (video) or an AAC frame (audio).
    pub data: Bytes,
    /// Decode timestamp in the movie `timescale` (derived from the RTMP timestamp,
    /// rebased so the session starts near 0 and never runs backwards).
    pub dts: u64,
    /// Presentation timestamp in `timescale` (`dts` + composition offset for B-frames).
    pub pts: u64,
    /// Sample duration in `timescale`.
    pub duration: u32,
    /// True for an IDR keyframe — a segment may only *start* on one of these.
    pub keyframe: bool,
}

/// Build the CMAF **init segment** (`ftyp` + `moov`) from the codec config (V3).
///
/// TODO(V3): write the boxes — `ftyp` (major brand `cmfc`/`iso6`), then `moov` with a
/// `mvhd`, a video `trak` (whose `stsd`→`avc1`→`avcC` carries `avc_decoder_config`,
/// with `width`/`height`), an audio `trak` (`stsd`→`mp4a`→`esds` from the AAC config),
/// and a `mvex`→`trex` per track (declaring these tracks are fragmented). Zero samples
/// live here. It must be **byte-stable** for a given config — that's the caching
/// contract. Reuse project 11's box writer.
pub fn build_init(cfg: &CodecConfig) -> Result<Bytes, AppError> {
    let _ = cfg;
    todo!("V3: build the CMAF init segment (ftyp + moov with avcC/esds, no samples)")
}

/// A stateful live fragmenter: samples are pushed as they arrive off the RTMP session,
/// and a **part** (`moof`+`mdat`) is cut on demand once ~`part_secs` has accumulated,
/// a **segment** boundary being a part that starts on a keyframe.
///
/// The timeline (`baseMediaDecodeTime` in each `tfdt`) must advance monotonically
/// across the whole live session — that is what makes the parts glue seamlessly.
pub struct Fragmenter {
    cfg: CodecConfig,
    /// Samples accumulated since the last cut (the only media held in memory).
    pending: Vec<Sample>,
    /// The decode time of the next fragment's first sample — the running timeline
    /// anchor written into `tfdt`.
    base_decode_time: u64,
}

impl Fragmenter {
    pub fn new(cfg: CodecConfig) -> Self {
        Self {
            cfg,
            pending: Vec::new(),
            base_decode_time: 0,
        }
    }

    /// Buffer one access unit for the fragment currently forming.
    pub fn push(&mut self, sample: Sample) {
        self.pending.push(sample);
    }

    /// The codec config this fragmenter was built with (for `build_init`).
    pub fn config(&self) -> &CodecConfig {
        &self.cfg
    }

    /// Cut the buffered samples into one fragment (`moof`+`mdat`) — an LL-HLS part (V3).
    ///
    /// TODO(V3): write a `moof` (`mfhd` sequence number, one `traf` per track with
    /// `tfhd`, a `tfdt` whose `baseMediaDecodeTime` == `self.base_decode_time`, and a
    /// `trun` listing each sample's size/duration/flags/composition-offset) followed by
    /// an `mdat` holding the sample bytes in order. Then advance `base_decode_time` by
    /// the fragment's total duration and clear `pending`. The returned `Bytes` is what
    /// the live window hands to every viewer — build it once. Keep A/V in sync and the
    /// timeline monotonic across cuts.
    pub fn cut_part(&mut self) -> Result<Bytes, AppError> {
        // `base_decode_time` and `pending` are the live-timeline state a real cut
        // consumes and advances; threaded so the intent is unmistakable.
        let _ = (&self.pending, self.base_decode_time);
        todo!("V3: cut the buffered samples into one moof+mdat fragment (a part)")
    }
}

#[cfg(test)]
mod tests {
    // TODO(V3): prove the packager:
    //   - `build_init` is byte-stable for a fixed config and `ffprobe`-parseable;
    //   - feeding captured access units and concatenating `init + parts` decodes with
    //     monotonic, gapless PTS (`fragments_decode_and_are_gapless`);
    //   - `baseMediaDecodeTime` advances by exactly each fragment's duration, never
    //     backwards, across many cuts (`baseMediaDecodeTime_is_monotonic`);
    //   - a long synthetic stream never grows `pending` without bound (the window /
    //     cut cadence keeps memory flat — `window_bounds_memory`).
}

File: projects/13-live-ingest/src/live_ingest/fmp4.py
"""V3 — Live fMP4 repackaging: rewrap H.264/AAC into CMAF, no re-encode.

Module: `src/live_ingest/fmp4.py`.

The codecs arriving over RTMP are already `<video>`-playable; the job is not to
transcode them but to **remux**: build a CMAF **init segment** (`ftyp` + `moov`
carrying the codec setup and zero samples) and a running series of
**fragments** (`moof` + `mdat`) on a monotonic MP4 timeline, cutting **parts**
(~200 ms) on demand and starting **segments** on IDR keyframes. H.264 arrives
over RTMP in AVCC framing — length-prefixed NALUs — which is exactly MP4's, so
`Sample.data` goes into `mdat` byte-for-byte.

This overlaps project 11's box writing (`vod_streaming/segment.py`,
`build_init_segment` / `build_media_segment`) — reuse the discipline. What is
new is doing it **live**: the timeline comes from RTMP message timestamps
(32-bit milliseconds, wrapping) rather than a finished sample table, and only
the current fragment's samples are ever held. `docs/02-live-fmp4-remuxing.md`
teaches the box layout and the timeline.

## Python details worth deciding up front

**Write boxes inside-out, or patch sizes.** Every box starts with its own
32-bit size, which you do not know until its children are written. Two clean
shapes: build each box as a function returning `bytes` (children first, then
`struct.pack(">I4s", 8 + len(body), b"moof") + body`), or write into one
`bytearray` and back-patch the size with `struct.pack_into`. The first is
easier to test box by box; the second allocates less. Measure before caring.

**`b"".join(parts)` beats `+=`.** Concatenating `mdat` sample bytes with `+=` on
`bytes` copies the growing buffer each time — quadratic in samples per part.
Collect the pieces in a list and join once.

**Byte-stability is a pure-function discipline.** `build_init` for a given
`CodecConfig` must return identical bytes every call — the `immutable` cache
header on `init.mp4` depends on it. No `time.time()` in `mvhd`, no dict
iteration over something unordered, nothing that is not a function of the config.

**Integers do not overflow — which is the trap.** Rust would have caught a
`tfdt` that exceeded 32 bits at the type level. Python's `int` just keeps
growing, and `struct.pack(">I", 2**32)` raises only when you finally write it.
`baseMediaDecodeTime` in version-1 `tfdt` is 64-bit; use it. And unwrapping
RTMP's 32-bit millisecond wrap is arithmetic you do yourself (`& 0xFFFFFFFF`
and a jump check), because nothing here will wrap for you.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from enum import StrEnum

from .flv import AudioSpecificConfig, AvcDecoderConfig

__all__ = [
    "VIDEO_TIMESCALE",
    "CodecConfig",
    "Fragmenter",
    "Sample",
    "TrackKind",
    "build_init",
]

VIDEO_TIMESCALE = 90_000
"""Ticks per second for the video track. At 30 fps a frame is exactly 3,000
ticks; in milliseconds it is 33.33… — the timescale exists so durations are exact.
The audio track conventionally uses its sample rate (an AAC frame is then
exactly 1,024 ticks)."""


class TrackKind(StrEnum):
    VIDEO = "video"
    AUDIO = "audio"


@dataclass(frozen=True, slots=True)
class CodecConfig:
    """Codec setup extracted from the sequence headers (V2), enough for an init segment.

    This is the *setup*, carried in `moov` — never per-frame media. Frozen and
    hashable-by-value, which is also what lets `build_init` be memoized on it.
    """

    video: AvcDecoderConfig
    audio: AudioSpecificConfig | None
    """`None` for a video-only publish (`ffmpeg ... -an`)."""


@dataclass(frozen=True, slots=True)
class Sample:
    """One access unit — a coded video frame or an AAC frame — to place in a fragment."""

    track: TrackKind
    data: bytes
    """Length-prefixed NALUs (video) or one raw AAC frame (audio), verbatim."""
    dts: int
    """Decode time in the track's timescale: derived from the RTMP timestamp,
    rebased so the session starts near zero, unwrapped so it never runs backwards."""
    pts: int
    """Presentation time in the track's timescale: `dts` plus the composition offset."""
    duration: int
    """In the track's timescale."""
    keyframe: bool
    """An IDR frame. A segment may only *start* on one."""


def build_init(config: CodecConfig) -> bytes:
    """Build the CMAF init segment, `ftyp` + `moov`, from the codec config (V3).

    TODO(V3): `ftyp` (major brand `iso6`, compatible `cmfc`), then `moov` with a
    `mvhd`, a video `trak` (`tkhd` with width/height; `stsd` → `avc1` → `avcC`
    holding `config.video.record` verbatim), an audio `trak` when there is audio
    (`stsd` → `mp4a` → `esds` wrapping `config.audio.record`), and `mvex` with
    one `trex` per track. Every sample table is present and **empty** — the
    samples live in fragments. Byte-stable for a given config.
    """
    raise NotImplementedError(
        "V3: build the CMAF init segment (ftyp + moov with avcC/esds, no samples)"
    )


class Fragmenter:
    """A stateful live fragmenter.

    Samples are pushed as they arrive off the session; `cut_part` wraps what has
    accumulated into one `moof` + `mdat` — an LL-HLS part. The decode time of
    each track's next fragment is the running anchor written into that track's
    `tfdt`, and it must advance by exactly each fragment's duration for the
    whole session. Nothing else is held: the memory bound is `pending`.
    """

    def __init__(self, config: CodecConfig) -> None:
        self.config = config
        self._pending: list[Sample] = []

    def push(self, sample: Sample) -> None:
        """Buffer one access unit for the fragment currently forming."""
        self._pending.append(sample)

    @property
    def pending(self) -> Sequence[Sample]:
        """The samples since the last cut — the only media this object holds."""
        return self._pending

    def cut_part(self) -> bytes:
        """Wrap the pending samples into one `moof` + `mdat` fragment (V3).

        TODO(V3): `mfhd` (a sequence number that increments per fragment), then
        per track a `traf`: `tfhd` (track id, defaults), `tfdt` (version 1,
        `baseMediaDecodeTime` = that track's running anchor), and `trun` (sample
        count, data offset into `mdat`, and each sample's
        duration/size/flags/composition offset — sync-sample flags on
        keyframes). Then `mdat` with the sample bytes in `trun` order. Advance
        each track's anchor by its fragment duration and clear `pending`.

        Where the part/segment cut *decisions* live — here, or in the session
        that calls this — is yours; the returned bytes are built once and served
        to every viewer from the live window.
        """
        raise NotImplementedError(
            "V3: cut the buffered samples into one moof+mdat fragment (a part)"
        )
